### PR TITLE
Fall back to sorting by name on ties

### PR DESCRIPTION
**Pull request by allisonkarlitskaya**
If we tie on the mtime or the file size then fall back to sorting based on name — it's always unique.

This should give us a fully-defined sort order for all directories.

### CODE REVIEW

File: src/files-card-body.jsx
Comment (cockpituous) [inline src/files-card-body.jsx:59]: This added line is not executed by any test.

File: src/files-card-body.jsx
Comment (cockpituous) [inline src/files-card-body.jsx:61]: This added line is not executed by any test.

**Comment by allisonkarlitskaya** [comment]
I think the debian fail is relevant, so probably don't merge this yet...

**Comment by jelly** [comment]
> I think the debian fail is relevant, so probably don't merge this yet...

Seems so! I'll download and take a look

**Comment by jelly** [comment]
So on debian testing the ordering on size is correct except for the first directory:

* eee
* Eee
* ddd

Should be

* ddd
* eee
* Eee

`ddd` is a symlink to `/tmp` so should be treated as a directory:
```
{
    "type": "lnk",
    "group": "root",
    "size": 4,
    "target": "/tmp",
    "mode": 511,
    "mtime": 1718701039.496,
    "user": "root",
    "name": "ddd",
    "to": "dir",
    "category": null
}
```

**Review by jelly — CHANGES_REQUESTED** [review]
Needs a fix for sorting on directories as they are compared by size when they are both a directory.

**Review by jelly — COMMENTED** [review]
Ironically this fails on typescript checks :-)

File: src/files-card-body.tsx
Comment (allisonkarlitskaya) [reply src/files-card-body.tsx:47]: That's what the check for `.type === "reg"` is for.  We completely ignore the size except for on regular files.

File: src/files-card-body.tsx
Comment (cockpituous) [inline src/files-card-body.tsx:48]: This added line is not executed by any test.

File: src/files-card-body.tsx
Comment (cockpituous) [inline src/files-card-body.tsx:56]: This added line is not executed by any test.

File: src/files-card-body.tsx
Comment (cockpituous) [inline src/files-card-body.tsx:58]: This added line is not executed by any test.

File: src/files-card-body.tsx
Comment (cockpituous) [inline src/files-card-body.tsx:64]: This added line is not executed by any test.

File: src/files-card-body.tsx
Comment (cockpituous) [inline src/files-card-body.tsx:450]: This added line is not executed by any test.

**Comment by allisonkarlitskaya** [comment]
The pixel test fails are kinda interesting.  Sometimes the `/` separators in the path bar get rendered a little bit thicker or thinner.  I wonder what's causing that...

https://cockpit-logs.us-east-1.linodeobjects.com/pull-535-2f0f2494-20240619-195556-fedora-40/log.html

image

File: src/files-card-body.tsx
Comment (cockpituous) [inline src/files-card-body.tsx:48]: This added line is not executed by any test.

File: src/files-card-body.tsx
Comment (cockpituous) [inline src/files-card-body.tsx:56]: This added line is not executed by any test.

File: src/files-card-body.tsx
Comment (cockpituous) [inline src/files-card-body.tsx:58]: This added line is not executed by any test.

File: src/files-card-body.tsx
Comment (cockpituous) [inline src/files-card-body.tsx:448]: This added line is not executed by any test.

**Comment by allisonkarlitskaya** [comment]
The pixel failures on `fedora-40` are also broken on main and fixed by #576.

File: src/fileActions.jsx
Comment (allisonkarlitskaya) [reply src/fileActions.jsx]: Okay.  I brought back `{ type: "divider" }` with a bit of discriminated union magic (and a bonus of some extra code in typescript for my efforts).

File: src/files-card-body.tsx
Comment (cockpituous) [inline src/files-card-body.tsx:49]: This added line is not executed by any test.

File: src/files-card-body.tsx
Comment (cockpituous) [inline src/files-card-body.tsx:57]: This added line is not executed by any test.

File: src/settings-dropdown.tsx
Comment (jelly) [inline src/settings-dropdown.tsx:57]: Note to self, this is basically the `KebabDropdown`...

File: src/menu.tsx
Comment (jelly) [inline src/menu.tsx:40]: > TypeScript introduced a new type never, which indicates the values that will never occur.

But uh, this will be set to a string at some point?

File: src/menu.tsx
Comment (jelly) [inline src/menu.tsx:4]: 2017? Sounds like a long time ago :)

File: src/menu.tsx
Comment (jelly) [inline src/menu.tsx:78]: At some point a huge function becomes unreadable in a list

**Review by jelly — APPROVED** [review]
Some of the commits could have been separate PRs, the menu changes will need another round of rebasing for feature PRs.

**Comment by jelly** [comment]
Needs some small fixes, so I retracted my PR.

File: src/menu.tsx
Comment (allisonkarlitskaya) [reply src/menu.tsx:40]: No.

This is basically a way of discriminating the union: either we have items that set `type` to the literal string `"divider"`, or it's forbidden from appearing at all (`never`).

**Review by jelly — APPROVED** [review]

**Comment by allisonkarlitskaya** [comment]
I opened https://github.com/cockpit-project/cockpit/issues/20635 for the arch failure — it's definitely not our issue, but it seems to be aggravated by the test case change.

I'm trying it again.  We can't really land this with a test that goes red 50% of the time, though...

**Comment by jelly** [comment]
So this needs a naughty now, as the Arch test always fails?

**Comment by allisonkarlitskaya** [comment]
> So this needs a naughty now, as the Arch test always fails?

Ya, It's a minor abuse of naughties, but from the standpoint of cockpit-files, `cockpit-bridge` *is* part of the set of system packages that we get from arch... The fact that it's only one distro where this happens also makes it feel more "right".  And even once we fix this in the bridge, we need to wait for it to filter in...

**Comment by allisonkarlitskaya** [comment]
`arch@bots#6532` went green so I consider this PR good to go once https://github.com/cockpit-project/bots/pull/6532 lands.

**Pull request merged**
